fcrepo-auth-webac/src/main/java/org/fcrepo/auth/webac/WebACRolesProvider.java

```diff
@@ -248,6 +249,12 @@ private FedoraResource getBaseVersion(final FedoraResource resource) {
                         effectiveRoles.computeIfAbsent(agent, key -> new HashSet<>())
                             .addAll(auth.getModes().stream().map(URI::toString).collect(toSet()));
                     });
+                auth.getAgentClasses().stream()
+                    .forEach(agentClass -> {
+                        // Use the AGENT_CLASS_SUFFIX to distinguish from agent roles
+                        effectiveRoles.computeIfAbsent(agentClass + AGENT_CLASS_SUFFIX, key -> new HashSet<>())
```

Comment:
I have used a Key suffix to distinguish agent roles from agentClass roles. I thought this will have less memory overhead. Would be better if a nested map is used instead where the agent roles and agentClass roles are on two separate HashMap and return a nested HashMap containing these two HashMaps?

Reply:
I don't think agents will have roles anymore, instead you make an `acl:agentGroup` which is just a list of `acl:agents`. 
I'd confirm with @peichman-umd and @dbernstein but I think the only user principal that will still use `acl:agentClass` is the public access `foaf:Agent`

Reply:
@whikloj That is correct, according to the [SOLID spec section on agents](https://github.com/fcrepo4/fcrepo4/pull/1324), the only recognized use of `acl:agentClass` is with `foaf:Agent` to indicate "Public Access (All Agents)".

Reply:
@whikloj Side opinion: this makes me wish the WebAC spec had just defined an individual (like `acl:everyone`) to mean "everyone", and just used `<> acl:agent acl:everyone`. I feel like `<> acl:agentClass foaf:Agent` may be a holdover from original plans to allow arbitrary agent classes, the same way that `acl:accessToClass` allows use of arbitrary classes on the protected resources.

Reply:
@peichman-umd you are probably correct. You could suggest that to the SOLID folks and see if they'd update their spec 😉

Reply:
@peichman-umd The target of `acl:agent` is a particular `foaf:Agent` (i.e. a Named Individual), so `<> acl:agent acl:everyone` would refer to one particular agent just as `<> acl:agent <http://example.com/me>` refers to one particular agent. If, on the other hand, you want to refer to a class of named individuals: for example "everyone", then you need something different than `acl:agent`, since you are referring to a "class" of named individuals rather than a particular named individual. Hence: `<> acl:agentClass foaf:Agent` or `<> acl:agentClass acl:AuthenticatedAgent` or `<> acl:agentClass ex:LocalSubDomainAgent`.

This isn't really a WebACL issue, it's an RDF issue.

Reply:
@whikloj Created another PR #1334 as an alternate implementation. I think it is a bit less complex and it is also more explicit in including the agentClass-Role mapping only when the agentClass is foaf:Agent.